docs/custom/error-handlers.mdx

```diff
@@ -0,0 +1,154 @@
+---
+id: error_handlers
+title: Custom error handlers
+---
+
+:::caution
+This AMF functionality is currently a work in progress. The following features are being developed:
+- Support for custom error handling in parsing, validation and emission stages
+- Javascript support
+:::
+
+In this documentation you'll learn what is an error handler in AMF and how to create one yourself with code examples.
+
+
+# What's an AMF error handler
+
+An error handler is a mechanism that lets you handle any errors that arise while using AMF.
+AMF allows you to define custom error handlers that you can customize and plug into AMF's resolution stage (and in the future, all stages).
+
+There are 3 types of errors in AMF, denoted by the `amf.core.validation.SeverityLevels` object:
+
+- **Info**
+  - This severity level points to something that may or may not be wrong
+  - Usually depends on user intent or use case
+- **Warning**
+  - This severity level indicates to the user that something is not correct, and should be changed
+  - Usually warns about future changes, using deprecated tools, or errors that will become violations the future
+- **Violation**
+  - This severity level indicates that something is definitely wrong
+  - Indication that something is not compliant with its specification
+  - This severity makes the AMF graph model invalid
+
+# Default error handling process
+AMF internally transforms each error into an `ValidationResult`, then collects all results generated in a `ValidationReport`
+that is returned after the validation stage, as seen in [previous examples](../AMF/using-amf/amf_validation#validation-examples).
+
+```scala title="amf.core.validation.core.ValidationReport"
+trait ValidationResult {
+  def message: Option[String]
+  def path: String
+  def sourceConstraintComponent: String
+  def focusNode: String
+  def severity: String
+  def sourceShape: String
+}
+
+trait ValidationReport {
+  def conforms: Boolean
+  def results: List[ValidationResult]
+}
+```
+## ValidationReport example
```

Comment:
## Example `ValidationReport`

For example, consider the following OAS 3.0 API:

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "test api",
    "version": "1"
  }
}
```
This API is invalid as it doesn't have the `paths:{}` object that is required by the
[OpenAPI 3.0 Specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#schema). In this case, AMF returns the following `ValidationReport`:

```text
Profile: OAS 3.0
Conforms? false
Number of results: 1
- Source: http://a.ml/vocabularies/amf/parser#mandatory-paths-property
  Message: 'paths' is mandatory in OAS spec
  Level: Violation
  Target: file://amf-client/shared/src/test/resources/validations/oas3/paths-property.json#/web-api
  Property:
  Position: None
  Location:
```
The report contains detailed error information, but with a custom error handler, you could implement additional behavior to:

- Stop execution when a certain type of error is found
- Process, modify or log incoming errors
- Run and check errors of a single stage (for example, parse an OAS 2.0 API and check if it has parsing errors)

To add custom functionality, you need to implement the provided **error handler interface**.